### ISSUE TITLE
When user type couple of first letters of components name for closing tag code completion does not appear

### ISSUE THREAD

**Issue by piotrzarzycki21** [prominic/Moonshine-IDE#57]
Code completion doesn't appear if user type couple of first letter for component in closing tag.

Gif which illustrate that problem:
![code_completion_closing_tags](https://user-images.githubusercontent.com/24554795/32232875-5866b01c-be59-11e7-84a6-bf3381745139.gif)

**Comment by piotrzarzycki21** [comment]
I will check first whether it is problem on the Moonshine sight.

**Comment by piotrzarzycki21** [comment]
I just checked and in described case above in the completion array which come from the language server I don't have at all components names. I see in that array only event names. I will assign this one to @joshtynjala for check.

![image](https://user-images.githubusercontent.com/24554795/32235222-2ad0aecc-be5f-11e7-8ea4-305c1c99c16b.png)

**Comment by joshtynjala** [comment]
I can confirm that this issue exists in vscode-nextgenas. I have not yet determined a way to solve it.

Unfortunately, the MXML code model in the Royale/FlexJS compiler is not as mature as the ActionScript code model, and I haven't yet found a way to detect that we're in an incomplete close tag that should match up with a previous open tag without a close tag.

This will need to remain as a known issue for now. I may revisit it again in the future. I'm removing the suggested milestone.

**Comment by piotrzarzycki21** [comment]
I think to avoid that issue would be implement in Moonshine auto closing tag mechanism. I remember that you have said this would be responsibility on IDE sight to make it happen.
I miss such thing and probably will take my hands on this one.

**Comment by joshtynjala** [comment]
I guess I just needed to sleep on it! In vscode-nextgenas, a closing tag will appear in the completion list now, if the last open tag is unclosed.

**Comment by piotrzarzycki21** [comment]
Sweet. Speaking of that commit. Is there any way that I could build myself code completion and use it in Moonshine for tests such changes ? If I go through your build instruction do I produce valid codecompletion.jar for Moonshine ?

**Comment by joshtynjala** [comment]
In *vscode-nextgenas_MoonshineModifications/src*, delete the *com* directory. Then, replace it with the latest source code from the language-server module in vscode-nextgenas.

https://github.com/BowlerHatLLC/vscode-nextgenas/tree/master/language-server/src/main/java

Inside *vscode-nextgenas_MoonshineModifications*, run the following command:

``` bash
ant deploy
```

This should work most of the time. If I've updated any dependencies in vscode-nextgenas, though, some of the JAR files won't be compatible.

I wish that *vscode-nextgenas_MoonshineModifications* was being built with Maven instead of Ant so that dependencies could be resolved automatically.

**Comment by piotrzarzycki21** [comment]
@joshtynjala Where I can find repository "vscode-nextgenas_MoonshineModifications" ?

**Comment by piotrzarzycki21** [comment]
@joshtynjala I just built with your instruction codecompletion.jar with latest changes. I can confirm that I see closing tag.

![image](https://user-images.githubusercontent.com/24554795/32771307-960ee992-c922-11e7-83db-65880234fc85.png)

Thanks for the instruction!

**Comment by joshtynjala** [comment]
Commit 38a63e61d4dfe67e15d5e5ee87c2c5cb4b279cc7 includes my fix for this. However, the closing tag seems to appear intermittently in the completion list. I confirmed that it's also intermittent in VSCode. I guess I may have solved it for a specific arrangement of XML tags, and it may not cover every possible situation.

I got as far as I could for now, I think. I may revisit this issue someday in the future.
